File: packages/service/lib/services/video.ts
import { AxiosInstance } from 'axios';
import queryString from 'query-string';
import { SubmitVideoDTO, VideoDTO } from '../dtos';
import { Video } from '../interfaces';
import { ActressService } from './actress';

export class VideoService {
  #axios: AxiosInstance;

  constructor(a: AxiosInstance) {
    this.#axios = a;
  }

  static converter(dto: VideoDTO): Video {
    return {
      id: dto.id,
      serial_number: dto.serial_number,
      cover_url: dto.cover_url,
      title: dto.title ?? undefined,
      chinese_title: dto.chinese_title ?? undefined,
      actresses: dto.actresses?.map(ActressService.converter),
      release_date: dto.release_date ?? undefined,
      video_url: dto.video_url ?? undefined,
      mosaic: dto.mosaic ?? undefined,
      tags: dto.tags ?? undefined,
      synopsis: dto.synopsis ?? undefined,
      created_at: dto.created_at,
      updated_at: dto.updated_at,
    };
  }

  list = async (params: { tags?: string[] }) => {
    const { data } = await this.#axios.get<VideoDTO[]>('/videos', {
      params,
      paramsSerializer: (params) => queryString.stringify(params),
    });
    return data.map(VideoService.converter);
  };

  paginate = async (req: { tags?: string[]; page?: number; size?: number }) => {
    const { data } = await this.#axios.post<{ data: VideoDTO[]; total: number }>('/videos/paginate', req);
    return { data: data.data.map(VideoService.converter), total: data.total };
  };

  create = async (dto: SubmitVideoDTO) => {
    const { data } = await this.#axios.post<VideoDTO>('/videos', dto);
    return VideoService.converter(data);
  };

  getById = async (videoId: number) => {
    const { data } = await this.#axios.get<VideoDTO>(`/videos/${videoId}`);
    return VideoService.converter(data);
  };

  update = async (videoId: number, dto: Partial<SubmitVideoDTO>) => {
    const { data } = await this.#axios.put<VideoDTO>(`/videos/${videoId}`, dto);
    return VideoService.converter(data);
  };

  delete = async (videoId: number) => {
    await this.#axios.delete(`/videos/${videoId}`);
  };
}

File: apps/backend/internal/app/server/dto/video.go
package dto

import (
	"encoding/json"
	"time"
)

type VideoDTO struct {
	ID           uint             `json:"id"`
	SerialNumber string           `json:"serial_number"`
	CoverUrl     string           `json:"cover_url"`
	Title        *string          `json:"title"`
	ChineseTitle *string          `json:"chinese_title"`
	Actresses    []*ActressDTO    `json:"actresses"`
	ReleaseDate  *time.Time       `json:"release_date"` // 发行日期
	VideoUrl     *string          `json:"video_url"`
	Mosaic       *bool            `json:"mosaic"`
	Tags         *json.RawMessage `json:"tags"`
	Synopsis     *string          `json:"synopsis"` // 概要
	CreatedAt    *time.Time       `json:"created_at"`
	UpdatedAt    *time.Time       `json:"updated_at"`
}

type CreateVideoDTO struct {
	SerialNumber string           `json:"serial_number" validate:"required"`
	CoverPath    string           `json:"cover_path" validate:"required"` // 封面在对象存储中桶内的路径
	Title        *string          `json:"title"`
	ChineseTitle *string          `json:"chinese_title"`
	Actresses    []*ActressDTO    `json:"actresses"`
	ReleaseDate  *time.Time       `json:"release_date"` // 发行日期
	VideoPath    *string          `json:"video_path"`   // 视频在对象存储中桶内的路径
	Mosaic       *bool            `json:"mosaic"`
	Tags         *json.RawMessage `json:"tags"`
	Synopsis     *string          `json:"synopsis"` // 概要
}

type UpdateVideoDTO struct {
	CreateVideoDTO
}
